### PR TITLE
Palette Testing & Grey Function

### LINKED ISSUE #3347: Larger color ranges (particularly gray scale)
Often I find myself using `bokeh` to plot some sort of image and overlay it with annotations. To do a reasonable job showing a grayscale image, I would really like to use more than a handful. I have tried using 256 grays, but this still feels limiting. Is there some possibility to have larger color ranges in `bokeh`? If there isn't some way to do this currently, what would be required to have this feature at some point?

### PR DESCRIPTION
**Pull request by HuntJSparra**
Supersedes #4696
Partially Resolves #3347 

- Adds grey/gray function similar to magma, plasma, and viridis
- Adds Greys10 & Greys11
- Adds testing for Bokeh palettes

### CODE REVIEW

File: bokeh/palettes.py [hunk 1]
```diff
@@ -486,14 +510,20 @@
...
-        """Return a palette of n equally spaced colors."""
+        if n>len(cmap): raise ValueError("Requested %(r)s colors, function can only return colors up to the base palette's length (%(l)s)" % {'r':n,'l':len(cmap)})
         return [cmap[int(math.floor(i))] for i in np.linspace(0, len(cmap)-1, num=n)]
     func.__name__ = name
     return func
 
-_cmaps = {name: _cmap_func_generator(name, cmap) for name, cmap in (('magma', Magma256),
+def grey_func(n):
+    cmap = Greys256
+    if n>256: raise ValueError("Hexidecimals support only up to 256 shades of grey")
```
Comment (bryevdv) [inline bokeh/palettes.py]: typo: Hexadecimals, but I would maybe just suggest "Gray scale only supports up to 256 shades"

**Comment by bryevdv** [comment]
LGTM will merge when tests are done.

**Comment by bryevdv** [comment]
@HuntJSparra so close :) there is a flake8 line length violation now:

```
bokeh/palettes.py:515:161: E501 line too long (163 > 160 characters)
```

**Comment by canavandl** [comment]
You may need to update: https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/palettes/palettes.coffee

File: bokeh/palettes.py [hunk 2]
```diff
@@ -486,14 +510,21 @@
...
     def func(n):
-        """Return a palette of n equally spaced colors."""
+        if n>len(cmap): raise ValueError("Requested %(r)s colors, function can only return colors up to the base palette's length (%(l)s)" % 
+            {'r':n,'l':len(cmap)})
         return [cmap[int(math.floor(i))] for i in np.linspace(0, len(cmap)-1, num=n)]
     func.__name__ = name
     return func
 
-_cmaps = {name: _cmap_func_generator(name, cmap) for name, cmap in (('magma', Magma256),
+def grey_func(n):
```
Comment (canavandl) [inline bokeh/palettes.py]: I'm sure I missed the discussion - but what's the logic for breaking this into a different function from `_linear_cmap_func_generator`?

File: bokeh/tests/test_palettes.py [hunk 3]
```diff
@@ -1,4 +1,440 @@
...
+    assert sorted(palettes, key=ret_ns_key) == sorted(pal.__palettes__, key=ret_ns_key)
+
+def test_palette_lengths():
+    palettes = [x for x in dir(pal) if isinstance(getattr(pal,x),list)]
+    # __palettes__ is also return and should not be tested
+    palettes = palettes[:-1]
+    for palstr in palettes:
+        assert len(getattr(pal,palstr)) == int(re.findall('\d+',palstr)[0])
+
+def test_cmap_generator_functions():
```
Comment (canavandl) [inline bokeh/tests/test_palettes.py]: While this "works", it's hard to read and will be a pain to update in the future. Generally, it's a good strategy to stay away from string manipulation if possible. I'd recommend something like:

```python
funcs = [x for x in dir(pal) if isinstance(getattr(pal, x), FunctionType)]
# First value is _cmap_func_generator and that should not be directly tested
funcs = funcs[1:]
cmap_functions = {viridis: Viridis256, gray: Grey256, ...}
# check that you're testing all of the cmap funcs (and that there aren't any extra)
assert len(set(funcs).symmetric_difference(cmap_functions.keys())) == 0
# check that they all create what's expected
for k, v in cmap_functions.items():
    assert k(256)==v
```

**Comment by HuntJSparra** [comment]
Will update coffee palettes first thing tomorrow

**Comment by HuntJSparra** [comment]
@canavandl Coffee is updated and everything should be ready to merge (hopefully). Thanks for looking over this and making sure I didn't break anything!

Regarding the separate grey_cmap function, @bryevdv wanted it separate because unlike other functions which can be updated to use (for example) a hypothetical Magma512 palette, until there is a better way to display and represent shades of grey there can only be 256 unique greys. I believe it is also separate so that if it fails due to that, debugging is easier.

File: bokehjs/src/coffee/palettes/palettes.coffee [hunk 4]
```diff
@@ -179,6 +179,31 @@ palettes = {
     Greys7      : [0x252525, 0x525252, 0x737373, 0x969696, 0xbdbdbd, 0xd9d9d9, 0xf7f7f7]
     Greys8      : [0x252525, 0x525252, 0x737373, 0x969696, 0xbdbdbd, 0xd9d9d9, 0xf0f0f0, 0xffffff]
     Greys9      : [0x000000, 0x252525, 0x525252, 0x737373, 0x969696, 0xbdbdbd, 0xd9d9d9, 0xf0f0f0, 0xffffff]
+    Greys10     = [0x000000, 0x1c1c1c, 0x383838, 0x555555, 0x717171, 0x8d8d8d, 0xaaaaaa, 0xc6c6c6, 0xe2e2e2, 0xffffff]
```
Comment (canavandl) [inline bokehjs/src/coffee/palettes/palettes.coffee]: need to be colons not equal signs

File: bokeh/tests/test_palettes.py [hunk 5]
```diff
@@ -1,4 +1,440 @@
...
+    funcs.remove('_linear_grey_cmap_func_generator')
+    cmap_functions = {viridis: Viridis256, gray: Greys256, grey: Greys256, magma: Magma256,
+        plasma: Plasma256, inferno: Inferno256}
+    # check that you're testing all of the cmap funcs (and that there aren't any extra)
+    keys = [y.__name__ for y in cmap_functions.keys()]
+    assert sorted(keys) == sorted(funcs)
+    # check that they all create what's expected
+    for k, v in cmap_functions.items():
+        assert k(256)==v
+
```
Comment (canavandl) [inline bokeh/tests/test_palettes.py]: I like this better - much more explicit and easy to read.

File: bokeh/tests/test_palettes.py [hunk 6]
```diff
@@ -1,4 +1,440 @@
...
+        'Greys8'      : ["#252525", "#525252", "#737373", "#969696", "#bdbdbd", "#d9d9d9", "#f0f0f0", "#ffffff"],
+        'Greys9'      : ["#000000", "#252525", "#525252", "#737373", "#969696", "#bdbdbd", "#d9d9d9", "#f0f0f0", "#ffffff"],
+        'Greys10'     : ['#000000', '#1c1c1c', '#383838', '#555555', '#717171', '#8d8d8d', '#aaaaaa', '#c6c6c6', '#e2e2e2', '#ffffff'],
+        'Greys11'     : ['#000000', '#191919', '#333333', '#4c4c4c', '#666666', '#7f7f7f', '#999999', '#b2b2b2', '#cccccc', '#e5e5e5', '#ffffff'],
+        'Greys256'    : [
+            "#000000", "#010101", "#020202", "#030303", "#040404", "#050505", "#060606", "#070707", "#080808", "#090909", "#0a0a0a", "#0b0b0b",
+            "#0c0c0c", "#0d0d0d", "#0e0e0e", "#0f0f0f", "#101010", "#111111", "#121212", "#131313", "#141414", "#151515", "#161616", "#171717",
+            "#181818", "#191919", "#1a1a1a", "#1b1b1b", "#1c1c1c", "#1d1d1d", "#1e1e1e", "#1f1f1f", "#202020", "#212121", "#222222", "#232323",
+            "#242424", "#252525", "$262626", "#272727", "#282828", "#292929", "#2a2a2a", "#2b2b2b", "#2c2c2c", "#2d2d2d", "#2e2e2e", "#2f2f2f",
```
Comment (canavandl) [inline bokeh/tests/test_palettes.py]: $262626?

File: bokeh/palettes.py [hunk 7]
```diff
@@ -486,14 +510,24 @@
...
+        if n>len(cmap): raise ValueError("Requested %(r)s colors, function can only return colors up to the base palette's length (%(l)s)" % 
+            {'r':n,'l':len(cmap)})
         return [cmap[int(math.floor(i))] for i in np.linspace(0, len(cmap)-1, num=n)]
     func.__name__ = name
     return func
 
-_cmaps = {name: _cmap_func_generator(name, cmap) for name, cmap in (('magma', Magma256),
+def _linear_grey_cmap_func_generator(name, cmap):
+    def func(n):
+        cmap = Greys256
```
Comment (canavandl) [inline bokeh/palettes.py]: overrides cmap arg in function definition. (last comment, I swear)

**Comment by canavandl** [comment]
LGTM - i'll merge when tests are green.

**Pull request merged**